src/modules/storefront/editor.tsx

```diff
@@ -0,0 +1,227 @@
+  if (has<'response'>('response', f)) {
+    return f.response;
+  } else {
+    return f;
+  }
+};
+
+export const reduceFieldData = (data: FieldData[]) =>
+  reduce<FieldData, Record<string, any>>(
+    (acc, data) => assocPath(data.name instanceof Array ? data.name : [data.name], data.value, acc),
+    {},
+    data
+  );
+
+export interface StorefrontEditorProps {
+  track: GoogleTracker;
+}
+
+export const validateSubdomainUniqueness = (
+  ar: ArweaveScope,
+  allowPubkey?: string
+): ((rule: RuleObject, subdomain: string | null | undefined) => Promise<void>) => {
+  return async (_, subdomain) => {
+    const storefront = await ar.storefront.find('holaplex:metadata:subdomain', subdomain ?? '');
+
+    if (isNil(storefront)) return;
+    if (allowPubkey && storefront.pubkey === allowPubkey) return;
+
+    throw new Error('The subdomain is already in use.  Please pick another.');
+  };
+};
+
+export const validateArweaveFunds = (
+  arweaveWalletAddress: string,
+  ar: ArweaveScope,
+  setShowARModal: (val: boolean) => void
+): ((rule: RuleObject, [file]: [FileInput?]) => Promise<void>) => {
+  return async (_, [file]) => {
+    if (isNil(file) || has('url', file)) return;
+
+    const canAfford =
+      arweaveWalletAddress && (await ar.wallet.canAfford(arweaveWalletAddress, file.size));
+
+    if (canAfford) return;
+
+    setShowARModal(true);
+
+    throw new Error('Not enough AR funds to cover the upload fee.');
+  };
+};
+
+export const submitCallback = ({
+  track,
+  router,
+  solana,
+  values,
+  setSubmitting,
+  successToast,
+  errorToast,
+  trackEvent,
+}: {
+  track: GoogleTracker;
+  router: NextRouter;
+  solana: Solana | undefined;
+  values: any;
+  setSubmitting: (val: boolean) => void;
+  successToast: (domain: string) => ReactChild;
+  errorToast: () => ReactChild;
+  trackEvent: string;
+}): (() => Promise<void>) => {
+  return async () => {
+    try {
+      setSubmitting(true);
+
+      const { theme, meta, subdomain } = values;
+      const domain = `${subdomain}.holaplex.com`;
+
+      const logo = popFile(theme.logo[0]);
+      const favicon = popFile(meta.favicon[0]);
+
+      await putStorefront({
+        solana,
+        storefront: {
+          theme: {
+            ...(theme as StorefrontTheme<undefined>),
+            logo: 'file' in logo ? undefined : logo,
+          },
+          meta: {
+            ...(meta as PageMetaData<undefined>),
+            favicon: 'file' in favicon ? undefined : favicon,
+          },
+          subdomain,
+        },
+        logo: logo.file,
+        favicon: favicon.file,
+        onProgress: (s) => console.log(s),
+        onError: (e) => console.log(e),
+      });
+
+      toast(successToast.bind(undefined, domain), { autoClose: 60000 });
+
+      router.push('/').then(() => {
+        track('storefront', trackEvent);
+        setSubmitting(false);
+      });
+    } catch {
+      setSubmitting(false);
+      toast.error(errorToast);
+    }
+  };
+};
```

Comment:
No need to refactor now but we discussed this function. What I think should still be managed within the view is setting the toast, calling setSumitting, and changing the route. This onSubmit helper should just post the storefront and error.

If the other items are removed can likely consolidate putStorefront into this method as well.